### PR TITLE
[SPARK-40562][SQL] Add `spark.sql.legacy.groupingIdWithAppendedUserGroupBy`

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to add a new legacy configuration to keep `grouping__id` value like the released Apache Spark 3.2 and 3.3.

Please note that this syntax is non-SQL standard and even Hive doesn't support it.

```SQL
hive> SELECT version();
OK
3.1.3 r4df4d75bf1e16fe0af75aad0b4179c34c07fc975
Time taken: 0.111 seconds, Fetched: 1 row(s)

hive> SELECT count(*), grouping__id from t GROUP BY a GROUPING SETS (b);
FAILED: SemanticException 1:63 [Error 10213]:
Grouping sets expression is not in GROUP BY key. Error encountered near token 'b'
```

### Why are the changes needed?

SPARK-40218 fixed a bug caused by SPARK-34932 (at Apache Spark 3.2.0). As a side effect, `grouping__id` values are changed. 

- Apache Spark 3.2.0, 3.2.1, 3.2.2, 3.3.0.
```scala
scala> sql("SELECT count(*), grouping__id from (VALUES (1,1,1),(2,2,2)) AS t(k1,k2,v) GROUP BY k1 GROUPING SETS (k2) ").show()
+--------+------------+
|count(1)|grouping__id|
+--------+------------+
|       1|           1|
|       1|           1|
+--------+------------+
```

- After SPARK-40218, Apache Spark 3.4.0, 3.3.1, 3.2.3
```scala
scala> sql("SELECT count(*), grouping__id from (VALUES (1,1,1),(2,2,2)) AS t(k1,k2,v) GROUP BY k1 GROUPING SETS (k2) ").show()
+--------+------------+
|count(1)|grouping__id|
+--------+------------+
|       1|           2|
|       1|           2|
+--------+------------+
```

- This PR (Apache Spark 3.4.0, 3.3.1, 3.2.3)
```scala
scala> sql("SELECT count(*), grouping__id from (VALUES (1,1,1),(2,2,2)) AS t(k1,k2,v) GROUP BY k1 GROUPING SETS (k2) ").show()
+--------+------------+
|count(1)|grouping__id|
+--------+------------+
|       1|           2|
|       1|           2|
+--------+------------+

scala> sql("set spark.sql.legacy.groupingIdWithAppendedUserGroupBy=true")
res1: org.apache.spark.sql.DataFrame = [key: string, value: string]scala> 

sql("SELECT count(*), grouping__id from (VALUES (1,1,1),(2,2,2)) AS t(k1,k2,v) GROUP BY k1 GROUPING SETS (k2) ").show()
+--------+------------+
|count(1)|grouping__id|
+--------+------------+
|       1|           1|
|       1|           1|
+--------+------------+ 
```

### Does this PR introduce _any_ user-facing change?

No, this simply added back the previous behavior by the legacy configuration.

### How was this patch tested?

Pass the CIs.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @cloud-fan , @wangyum , @viirya , @huaxingao , @sunchao

**Comment by dongjoon-hyun** [comment]
@cloud-fan As you wrote in the PR description (https://github.com/apache/spark/pull/32022), it's not in the SQL standard, is it?
> GROUP BY ... GROUPING SETS (...) is a weird SQL syntax we copied from Hive. It's not in the SQL standard or any other mainstream databases.

**Comment by viirya** [comment]
Is the grouping id values generated in Spark 3.2.0 same as previous versions (e.g. Spark 3.1.x and earlier)?

**Comment by dongjoon-hyun** [comment]
Before SPARK-34932, Apache Spark fails at `GROUP BY a GROUPING SETS(a, b)` and forced users to also put b after GROUP BY. SPARK-34932 allows it by working like `GROUP BY GROUPING SETS(a, b)`, @viirya .

**Comment by thiyaga** [comment]
We use grouping sets on our queries and rely on `grouping__id` to use as an identifier to query the data for respective group. If we use `grouping__id` directly, it will be prone to change if grouping set changes (for e.g. adding new grouping set/ adding new column to existing grouping set). Any grouping id change will make things even more complex when consuming this data directly from reporting tools like Tableau . We need to do the one of the following options to mitigate the changing `grouping__id`

1. Either we need to transform the `grouping__id` to something that won't be impacted when the grouping set changes and deterministic (for e.g convert `grouping__id` to `group_name`)
2. Have some sort of logical DB view which will handle the transformation at runtime (for e.g. using CASE WHEN)

In essence, we always have dependency with `grouping__id` when grouping sets are used in our query. Any change in the grouping id generation will have immediate impact. This new parameter will help us to use the legacy logic.

**Comment by dongjoon-hyun** [comment]
Thank you for your feedback, @thiyaga .

**Comment by cloud-fan** [comment]
> it's not in the SQL standard

Yea, but since we copied it from Hive, I think the result should match Hive as well. Sorry I didn't realize there is a result change when doing the bug fix PR. Can we try with Hive to make sure we return the correct result today?

**Review by cloud-fan — APPROVED** [review]

**Review by viirya — APPROVED** [review]
+1 for a legacy config.

I think we should not break an existing 3.2.0 behavior in later 3.2.x releases. `grouping_id` might be used by end-users and such change is unexpected and breaking.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -3574,6 +3574,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+   val LEGACY_GROUPING_ID_WITH_APPENDED_USER_GROUPBY =
+    buildConf("spark.sql.legacy.groupingIdWithAppendedUserGroupBy")
+      .internal()
+      .doc("When true, grouping_id() returns values based on grouping set columns plus " +
+        "user-given group-by expressions order.")
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3581]: Do we need to mention when the behavior of grouping_id has changed?

**Review by huaxingao — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you, @cloud-fan , @viirya , @huaxingao . Yes, as Wenchen shared, this is really Spark-specific syntax now. Let me add that to PR description.
```
hive> SELECT version();
OK
3.1.3 r4df4d75bf1e16fe0af75aad0b4179c34c07fc975
Time taken: 0.111 seconds, Fetched: 1 row(s)

hive> SELECT count(*), grouping__id from t GROUP BY a GROUPING SETS (b);
FAILED: SemanticException 1:63 [Error 10213]: Grouping sets expression is not in GROUP BY key. Error encountered near token 'b'
```

I'll revise the config according to @viirya 's comment and will add a migration guide for this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I added the old versions, `3.2.0, 3.2.1, 3.2.2, and 3.3.0`.

**Comment by viirya** [comment]
Thank you @dongjoon-hyun.

**Review by zhengruifeng — APPROVED** [review]
LGTM

**Comment by dongjoon-hyun** [comment]
Thank you again, @cloud-fan , @viirya , @thiyaga,  @huaxingao , @zhengruifeng . 
Since the last commit is about doc, I'll merge this.

Merged to master/3.3/3.2.

cc @wangyum , too

**Pull request closed without merging**